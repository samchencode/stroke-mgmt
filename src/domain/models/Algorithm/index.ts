export type { Algorithm } from '@/domain/models/Algorithm/Algorithm';
export type { AlgorithmVisitor } from '@/domain/models/Algorithm/AlgorithmVisitor';
export { TextAlgorithm } from '@/domain/models/Algorithm/TextAlgorithm';
export { ScoredAlgorithm } from '@/domain/models/Algorithm/ScoredAlgorithm';
export { AlgorithmInfo } from '@/domain/models/Algorithm/AlgorithmInfo';
export { AlgorithmId } from '@/domain/models/Algorithm/AlgorithmId';
export { Outcome } from '@/domain/models/Algorithm/Outcome';
export type { AlgorithmRepository } from '@/domain/models/Algorithm/ports/AlgorithmRepository';
export type { AlgorithmRenderer } from '@/domain/models/Algorithm/ports/AlgorithmRenderer';
export {
  GreaterThanCriterion,
  LessThanCriterion,
} from '@/domain/models/Algorithm/Criterion';
export {
  Switch,
  YesNoSwitch,
  SwitchId,
} from '@/domain/models/Algorithm/Switch';
export { RenderedAlgorithm } from '@/domain/models/Algorithm/RenderedAlgorithm';
export { RenderedAlgorithmCollection } from '@/domain/models/Algorithm/RenderedAlgorithmCollection';
