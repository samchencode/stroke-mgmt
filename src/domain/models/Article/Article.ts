import type { ArticleId } from '@/domain/models/Article/ArticleId';
import type { BaseDesignation } from '@/domain/models/Article/Designation/BaseDesignation';
import type { Image } from '@/domain/models/Image';
import type { Tag } from '@/domain/models/Tag';

type ArticleParams = {
  id: ArticleId;
  title: string;
  html: string;
  summary?: string;
  designation: BaseDesignation;
  thumbnail: Image;
  shouldShowOnHomeScreen: boolean;
  tags?: Tag[];
};

class Article {
  private id: ArticleId;

  private title: string;

  private html: string;

  private summary?: string;

  private desigation: BaseDesignation;

  private thumbnail: Image;

  private tags: Map<string, Tag>;

  private shouldShowOnHomeScreen: boolean;

  constructor({
    id,
    title,
    html,
    designation,
    thumbnail,
    summary,
    tags = [],
    shouldShowOnHomeScreen,
  }: ArticleParams) {
    this.id = id;
    this.title = title;
    this.html = html;
    this.desigation = designation;
    this.thumbnail = thumbnail;
    this.summary = summary;
    this.tags = new Map(tags.map((t) => [t.getName(), t] as const));
    this.shouldShowOnHomeScreen = shouldShowOnHomeScreen;
  }

  getId() {
    return this.id;
  }

  getTitle() {
    return this.title;
  }

  getHtml() {
    return this.html;
  }

  getThumbnail() {
    return this.thumbnail;
  }

  getSummary(sanitizeHtml: (html: string) => string): string {
    if (this.summary) return this.summary;
    const sanitizedBody = sanitizeHtml(this.html);
    return sanitizedBody.replace(/\s+/g, ' ').trim().slice(0, 100);
  }

  getDesignation() {
    return this.desigation;
  }

  getTags() {
    return [...this.tags.values()];
  }

  hasTag(tag: Tag): boolean {
    return this.tags.has(tag.getName());
  }

  getshouldShowOnHomeScreen() {
    return this.shouldShowOnHomeScreen;
  }

  is(v: Article) {
    return this.id.is(v.getId());
  }
}

export { Article };
