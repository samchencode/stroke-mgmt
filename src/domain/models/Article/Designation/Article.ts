import { BaseDesignation } from '@/domain/models/Article/Designation/BaseDesignation';

class Article extends BaseDesignation {
  readonly type = 'Article';
}

export { Article };
