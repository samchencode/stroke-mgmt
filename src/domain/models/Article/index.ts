export { ArticleId } from '@/domain/models/Article/ArticleId';
export { Designation } from '@/domain/models/Article/Designation';
export type { BaseDesignation } from '@/domain/models/Article/Designation';
export { Article } from '@/domain/models/Article/Article';
export type { ArticleRenderer } from '@/domain/models/Article/ports/ArticleRenderer';
export type { ArticleRepository } from '@/domain/models/Article/ports/ArticleRepository';
export { ArticleMetadata } from '@/domain/models/Article/ArticleMetadata';
export type { CachedArticleRepository } from '@/domain/models/Article/ports/CachedArticleRepository';
export { ArticleNotFoundError } from '@/domain/models/Article/ports/ArticleRepository';
export { CachedArticleNotFoundError } from '@/domain/models/Article/ports/CachedArticleRepository';
export { NullArticle } from '@/domain/models/Article/NullArticle';
