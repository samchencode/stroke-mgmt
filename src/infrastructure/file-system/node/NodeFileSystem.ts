import fs from 'fs';
import path from 'path';
import type { FileSystem } from '@/infrastructure/file-system/FileSystem';

const readUtf8File = (p: string) =>
  new Promise<string>((resolve, reject) => {
    fs.readFile(p, 'utf-8', (err, data) => (err ? reject(err) : resolve(data)));
  });

class NodeFileSystem implements FileSystem<'jest'> {
  async getAssetAsString(aliasedFilePath: string): Promise<string> {
    const filePath = aliasedFilePath.replace(
      /^@/,
      path.resolve(__dirname, '../../..')
    );
    return readUtf8File(filePath);
  }
}

export { NodeFileSystem };
