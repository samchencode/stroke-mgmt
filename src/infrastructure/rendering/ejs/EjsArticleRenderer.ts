import ejs from 'ejs';
import type { Article } from '@/domain/models/Article';
import type { ArticleRenderer } from '@/domain/models/Article/ports/ArticleRenderer';
import type { FileSystem } from '@/infrastructure/file-system/FileSystem';

class EjsArticleRenderer implements ArticleRenderer {
  ready: Promise<void>;

  fs: FileSystem;

  articleTemplate: ejs.TemplateFunction | null = null;

  disclaimerTemplate: ejs.TemplateFunction | null = null;

  constructor(fileSystem: FileSystem) {
    this.fs = fileSystem;
    this.ready = new Promise((s) => {
      this.init().then(s);
    });
  }

  private async init(): Promise<void> {
    const articleTemplate = await this.fs.getAssetAsString(
      require('@/infrastructure/rendering/ejs/templates/article.ejs')
    );
    this.articleTemplate = ejs.compile(articleTemplate);
    const disclaimerTemplate = await this.fs.getAssetAsString(
      require('@/infrastructure/rendering/ejs/templates/disclaimer.ejs')
    );
    this.disclaimerTemplate = ejs.compile(disclaimerTemplate);
  }

  async renderArticle(article: Article): Promise<string> {
    await this.ready;
    if (!this.articleTemplate) throw new Error('Article template not found');
    return this.articleTemplate({
      title: article.getTitle(),
      body: article.getHtml(),
    });
  }

  async renderDisclaimer(article: Article): Promise<string> {
    await this.ready;
    if (!this.disclaimerTemplate)
      throw new Error('Disclaimer template not found');
    return this.disclaimerTemplate({
      title: article.getTitle(),
      body: article.getHtml(),
    });
  }

  async renderStrokeFacts(article: Article): Promise<string> {
    return this.renderArticle(article);
  }

  async renderStrokeSigns(article: Article): Promise<string> {
    return this.renderArticle(article);
  }
}

export { EjsArticleRenderer };
