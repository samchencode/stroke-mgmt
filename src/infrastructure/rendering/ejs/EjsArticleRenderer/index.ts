export { EjsArticleRenderer } from '@/infrastructure/rendering/ejs/EjsArticleRenderer/EjsArticleRenderer';
