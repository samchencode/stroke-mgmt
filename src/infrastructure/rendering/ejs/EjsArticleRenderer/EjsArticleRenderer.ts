import ejs from 'ejs';
import type { Article } from '@/domain/models/Article';
import type { ArticleRenderer } from '@/domain/models/Article/ports/ArticleRenderer';
import type { FileSystem } from '@/infrastructure/file-system/FileSystem';

class EjsArticleRenderer implements ArticleRenderer {
  ready: Promise<void>;

  fs: FileSystem<unknown>;

  articleTemplate: ejs.TemplateFunction | null = null;

  disclaimerTemplate: ejs.TemplateFunction | null = null;

  constructor(fileSystem: FileSystem<unknown>) {
    this.fs = fileSystem;
    this.ready = new Promise((s) => {
      this.init().then(s);
    });
  }

  private async init(): Promise<void> {
    const articleTemplate = await this.fs.getAssetAsString(
      require('@/infrastructure/rendering/ejs/EjsArticleRenderer/article.ejs')
    );
    this.articleTemplate = ejs.compile(articleTemplate);
    const disclaimerTemplate = await this.fs.getAssetAsString(
      require('@/infrastructure/rendering/ejs/EjsArticleRenderer/disclaimer.ejs')
    );
    this.disclaimerTemplate = ejs.compile(disclaimerTemplate);
  }

  async renderArticle(article: Article): Promise<string> {
    await this.ready;
    if (!this.articleTemplate) throw new Error('Article template not found');
    return this.articleTemplate({
      title: article.getTitle(),
      body: article.getHtml(),
    });
  }

  async renderDisclaimer(article: Article): Promise<string> {
    await this.ready;
    if (!this.disclaimerTemplate)
      throw new Error('Disclaimer template not found');
    return this.disclaimerTemplate({
      title: article.getTitle(),
      body: article.getHtml(),
    });
  }

  async renderStrokeFacts(article: Article): Promise<string> {
    return this.renderArticle(article);
  }

  async renderStrokeSigns(article: Article): Promise<string> {
    return this.renderArticle(article);
  }
}

export { EjsArticleRenderer };
