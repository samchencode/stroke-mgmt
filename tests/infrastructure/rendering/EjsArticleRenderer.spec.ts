import { Article, ArticleId, Designation } from '@/domain/models/Article';
import { NodeFileSystem } from '@/infrastructure/file-system/node/NodeFileSystem';
import { EjsArticleRenderer } from '@/infrastructure/rendering/ejs/EjsArticleRenderer';

describe('EjsArticleRenderer', () => {
  let fs: NodeFileSystem;

  beforeAll(() => {
    jest.resetModules();
    jest.doMock(
      '@/infrastructure/rendering/ejs/EjsArticleRenderer/article.ejs',
      () => '@/infrastructure/rendering/ejs/EjsArticleRenderer/article.ejs'
    );
    jest.doMock(
      '@/infrastructure/rendering/ejs/EjsArticleRenderer/disclaimer.ejs',
      () => '@/infrastructure/rendering/ejs/EjsArticleRenderer/disclaimer.ejs'
    );

    fs = new NodeFileSystem();
  });

  describe('Instantiation', () => {
    it('should be created given file system with template', () => {
      const create = () => new EjsArticleRenderer(fs);
      expect(create).not.toThrowError();
    });
  });

  describe('Behavior', () => {
    it('should render article title and html', async () => {
      const article = new Article({
        id: new ArticleId('0'),
        title: 'hello world',
        html: '<h1>foo bar</h1>',
        designation: Designation.ARTICLE,
      });

      const renderer = new EjsArticleRenderer(fs);
      const html = await renderer.renderArticle(article);
      expect(html).toMatchSnapshot();
    });

    it('should render disclaimer html', async () => {
      const article = new Article({
        id: new ArticleId('1'),
        title: 'hello world',
        html: '<h1>foo bar</h1>',
        designation: Designation.DISCLAIMER,
      });

      const renderer = new EjsArticleRenderer(fs);
      const html = await renderer.renderDisclaimer(article);
      expect(html).toMatchSnapshot();
    });
  });
});
