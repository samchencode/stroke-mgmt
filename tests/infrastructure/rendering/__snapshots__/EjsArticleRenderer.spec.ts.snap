// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EjsArticleRenderer Behavior should render article title and html 1`] = `
"<!DOCTYPE html>
<html lang="en">

<head>
  <meta charset="UTF-8">
  <meta http-equiv="X-UA-Compatible" content="IE=edge">
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  <title>hello world</title>
  <style>
    * {
      font-family: Arial, Helvetica, sans-serif;
    }
  </style>
</head>

<body>
  <h1>foo bar</h1>
</body>

</html>"
`;

exports[`EjsArticleRenderer Behavior should render disclaimer html 1`] = `
"<!DOCTYPE html>
<html lang="en">

<head>
  <meta charset="UTF-8">
  <meta http-equiv="X-UA-Compatible" content="IE=edge">
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  <title>hello world</title>
  <style>
    * {
      margin: 0;
      padding: 0;
      box-sizing: border-box;
      font-family: Arial, Helvetica, sans-serif;
    }

    h1 {
      margin-bottom: 16px;
    }

    p {
      text-align: left;
    }
  </style>
</head>

<body>
  <h1>foo bar</h1>

  <script>
    var height = document.body.offsetHeight;
    var width = document.body.offsetWidth;
    var dims = {
      height: height,
      width: width
    }
    window.ReactNativeWebView.postMessage(JSON.stringify(dims))
  </script>
</body>

</html>"
`;
